core/Pos/Core/Genesis.hs

```diff
@@ -119,19 +119,24 @@ generateHdwGenesisSecretKey =
 -- related 'bootDustThreshold'. Function splits coin on chunks of
 -- @weightsSum@ and distributes it over boot stakeholder. Remainder is
 -- assigned randomly (among boot stakeholders) based on hash of the coin.
+--
+-- If coin is lower than 'bootDustThreshold' then this function
+-- distributes coins among first stakeholders in the list according to
+-- their weights.
 genesisSplitBoot ::
-       GenesisWStakeholders -> Coin -> Either Text [(StakeholderId, Coin)]
+       GenesisWStakeholders -> Coin -> [(StakeholderId, Coin)]
 genesisSplitBoot g@(GenesisWStakeholders bootWHolders) c
-    | cval <= 0 =
-      Left $ "sendMoney#splitBoot: cval <= 0: " <> show cval
     | c < bootDustThreshold g =
-      Left $
-      sformat ("Can't spend "%build%" coins: amount is too small for boot era, "%
-               "threshold is "%build%" in respect to boot stakeholders set "%build)
-              c
-              (bootDustThreshold g)
-              g
-    | otherwise = Right $ bootHolders `zip` stakeCoins
+          snd $
+          foldr (\(s,w) r@(totalSum, res) ->
+                   if totalSum >= cval then r
+                   else let w' = (fromIntegral w :: Word64)
+                            toInclude = fromIntegral $ min w' (totalSum - w')
```

Comment:
1. `fromIntegral` is probably redundant here.
2. I don't like that you abuse signed overflow here.
3. I don't understand how it works. Suppose stakeholders have weights 7, 2, 3. `cval` is 10.
First step: min 7 (-7) = 7 (overflow abuse).
Second step: min 2 5 = 2
Third step: min 3 6 = 3.

7 + 2 + 3 = 12 > 10.
Where am I wrong?